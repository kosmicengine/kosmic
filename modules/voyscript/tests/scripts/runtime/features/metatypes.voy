class MyClass:
	const TEST = 10

enum MyEnum {A, B, C}

const Other = preload("./metatypes.notest.voy")

var test_native := JSON
var test_script := Other
var test_class := MyClass
var test_enum := MyEnum

func check_voyscript_native_class(value: Variant) -> void:
	print(var_to_str(value).get_slice(",", 0).trim_prefix("Object("))

func check_voyscript(value: VoyScript) -> void:
	print(value.get_class())

func check_enum(value: Dictionary) -> void:
	print(value)

func test():
	for property in get_property_list():
		if str(property.name).begins_with("test_"):
			print(Utils.get_property_signature(property))

	print("---")
	check_voyscript_native_class(test_native)
	check_voyscript(test_script)
	check_voyscript(test_class)
	check_enum(test_enum)

	print("---")
	print(test_native.stringify([]))
	print(test_script.TEST)
	print(test_class.TEST)
	print(test_enum.keys())

	print("---")
	# Some users add unnecessary type hints to `const`-`preload`, which removes metatypes.
	# For **constant** `VoyScript` we still check the class members, despite the wider type.
	const ScriptNoMeta: VoyScript = Other
	const ClassNoMeta: VoyScript = MyClass
	var a := ScriptNoMeta.TEST
	var b := ClassNoMeta.TEST
	print(a)
	print(b)
