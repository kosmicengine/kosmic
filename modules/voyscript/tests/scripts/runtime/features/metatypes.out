VOYTEST_OK
var test_native: VoyScriptNativeClass
var test_script: VoyScript
var test_class: VoyScript
var test_enum: Dictionary
---
VoyScriptNativeClass
VoyScript
VoyScript
{ "A": 0, "B": 1, "C": 2 }
---
[]
100
10
["A", "B", "C"]
---
100
10
